### PR TITLE
feat(frontend): service util for rest requests

### PR DESCRIPTION
**Pull request by AntonioVentilii**
# Motivation

The Etherscan API has a maximum [calls/second limit](https://docs.etherscan.io/support/rate-limits) and with the unified transaction we are always breaking it.

There are a few ways to handle/solve this issue, but, as first important guard, we would like to create a way to handle retries in our requests to the Rest API. Right now a request would fail at the first try (if it fails).

So, we create a service handler (similar to `queryAndUpdate`) that accepts a request and handles success, errors and retries.

# Tests

Provided some tests.

### CODE REVIEW

File: src/frontend/src/lib/services/rest.services.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		} catch (error) {
+			retryCount++;
+
+			if (retryCount > maxRetries) {
+				console.error(`Max retries reached. Error:`, error);
+				return onError?.(error as E);
+			}
+
+			await onRetry?.({ error: error as E, retryCount });
+			console.warn(`Request attempt ${retryCount} failed. Retrying...`);
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/rest.services.ts:34]: same we need this console.warn?

File: src/frontend/src/lib/services/rest.services.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	onRetry,
+	maxRetries = 10
+}: RestRequestParams<R, E, S>): Promise<S | undefined> => {
+	let retryCount = 0;
+
+	while (retryCount <= maxRetries) {
+		try {
+			const response = await request();
+			return onSuccess(response);
+		} catch (error) {
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/rest.services.ts:25]: `error: unknown`

File: src/frontend/src/lib/services/rest.services.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ *
+ */
+export const restRequest = async <R, E, S>({
+	request,
+	onSuccess,
+	onError,
+	onRetry,
+	maxRetries = 10
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/rest.services.ts:17]: Do we want 10 as default? I imagined something smallerin a first time, like 3

File: src/frontend/src/tests/lib/services/rest.services.spec.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+		// we mock console.error and console.warn just to avoid unnecessary logs while running the tests
+		vi.spyOn(console, 'error').mockImplementation(() => undefined);
+		vi.spyOn(console, 'warn').mockImplementation(() => undefined);
+
+		beforeEach(() => {
+			vi.clearAllMocks();
+		});
+
+		it('should call onSuccess when the request succeeds on the first try', async () => {
```
Comment (peterpeterparker) [inline src/frontend/src/tests/lib/services/rest.services.spec.ts:21]: If I'm not mistaken, we need another test to assert the call succeed after first but before max.

File: src/frontend/src/lib/services/rest.services.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+export interface RestRequestParams<R, E = unknown, S = void> {
```
[the remark this answers is not in the document]
Comment (AntonioVentilii) [reply src/frontend/src/lib/services/rest.services.ts:1]: yes, makes sense indeed

File: src/frontend/src/lib/services/rest.services.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	request: () => Promise<R>;
+	onSuccess: (response: R) => S | undefined;
+	onError?: (error: E) => S | undefined;
+	onRetry?: (options: { error: E; retryCount: number }) => Promise<void>;
+	maxRetries?: number;
+}
+
+/**
```
[the remark this answers is not in the document]
Comment (AntonioVentilii) [reply src/frontend/src/lib/services/rest.services.ts:9]: ops! my bad! included

File: src/frontend/src/lib/services/rest.services.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+export const restRequest = async <R, E, S>({
+	request,
+	onSuccess,
+	onError,
+	onRetry,
+	maxRetries = 10
+}: RestRequestParams<R, E, S>): Promise<S | undefined> => {
+	let retryCount = 0;
+
+	while (retryCount <= maxRetries) {
```
[the remark this answers is not in the document]
Comment (AntonioVentilii) [reply src/frontend/src/lib/services/rest.services.ts:28]: as discussed offline, I will implement that pattern in https://github.com/dfinity/oisy-wallet/pull/3644

File: src/frontend/src/lib/services/rest.services.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+		try {
+			const response = await request();
+			return onSuccess(response);
+		} catch (error) {
+			retryCount++;
+
+			if (retryCount > maxRetries) {
+				console.error(`Max retries reached. Error:`, error);
```
[the remark this answers is not in the document]
Comment (AntonioVentilii) [reply src/frontend/src/lib/services/rest.services.ts:36]: in `updateAndQuery` we print it, but not needed no; i can remove it?

File: src/frontend/src/lib/services/rest.services.ts (hunk 3 above)
Comment (AntonioVentilii) [reply src/frontend/src/lib/services/rest.services.ts:24]: ah right! yes, totally random eheh

File: src/frontend/src/tests/lib/services/rest.services.spec.ts (hunk 4 above)
Comment (AntonioVentilii) [reply src/frontend/src/tests/lib/services/rest.services.spec.ts:21]: i created one that is `should succeed after retries if a later attempt is successful`, is that ok? or something else?

File: src/frontend/src/lib/services/rest.services.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/lib/services/rest.services.ts:36]: Ah then all good

File: src/frontend/src/tests/lib/services/rest.services.spec.ts (hunk 4 above)
Comment (peterpeterparker) [reply src/frontend/src/tests/lib/services/rest.services.spec.ts:21]: all good

**Review by peterpeterparker — APPROVED** [review]
LGTm, thx

**Comment by AntonioVentilii** [comment]
[OISY-130]

**Pull request merged**